File: server/src/main/java/object/Room.java
package object;

import auth.UserConnection;
import game.GameSettings;
import org.w3c.dom.Document;
import server.EntropyServer;
import util.*;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Server-side version of a Room
 */
public final class Room
{
	private boolean isCopy = false;
	
	private ExtendedConcurrentHashMap<Integer, String> hmPlayerByPlayerNumber = new ExtendedConcurrentHashMap<>();
	private ConcurrentHashMap<Integer, String> hmFormerPlayerByPlayerNumber = new ConcurrentHashMap<>();
	private ArrayList<OnlineMessage> chatHistory = new ArrayList<>();
	private List<String> currentPlayers = new ArrayList<>();
	private List<String> observers = new ArrayList<>();
	private GameWrapper previousGame = null;
	private GameWrapper currentGame = null;

	private String name;
	private GameSettings settings;
	private int capacity;
	private EntropyServer server;

	public Room(String name, GameSettings settings, int capacity, EntropyServer server)
	{
		this.name = name;
		this.settings = settings;
		this.capacity = capacity;
		this.server = server;
	}
	
	public boolean isFull()
	{
		return currentPlayers.size() == capacity;
	}
	public boolean isEmpty()
	{
		return currentPlayers.isEmpty()
		  && observers.isEmpty();
	}
	
	public void clearChatIfEmpty()
	{
		synchronized (this)
		{
			if (!isEmpty())
			{
				return;
			}
			
			chatHistory.clear();
		}
	}
	
	public int addToCurrentPlayers(String username, int playerNumber)
	{
		synchronized (this)
		{
			String existingUsername = hmPlayerByPlayerNumber.get(playerNumber);
			if (existingUsername != null)
			{
				Debug.append(username + " tried to join room " + name + " as player " + playerNumber
							+ " but the space was taken by " + existingUsername);
				return -1;
			}
			
			if (hmPlayerByPlayerNumber.containsValue(username))
			{
				Debug.append(username + " tried to join room " + name + " twice!");
				return -1;
			}
			
			currentPlayers.add(username);
			hmPlayerByPlayerNumber.put(playerNumber, username);
			observers.remove(username);
			
			notifyAllPlayersOfPlayerChange(username, false);
			ServerGlobals.lobbyService.lobbyChanged();
			return playerNumber;
		}
	}
	
	public void removePlayer(String username, boolean fireLobbyChanged)
	{
		for (int playerNumber=0; playerNumber<capacity; playerNumber++)
		{
			String user = hmPlayerByPlayerNumber.get(playerNumber);
			if (user != null 
			  && username.equals(user))
			{
				hmPlayerByPlayerNumber.remove(playerNumber);
				hmFormerPlayerByPlayerNumber.put(playerNumber, user);
				
				//Notify everyone in the room that this player has left. Block on this.
				notifyAllPlayersOfPlayerChange(username, true);
				
				//The game has not started
				if (currentGame.getGameStartMillis() == -1)
				{
					//Unset the countdown if it's going, reset current capacity and get out of this madness
					currentGame.setCountdownStartMillis(-1);
					resetCurrentPlayers(fireLobbyChanged);
					return;
				}
				
				//There is a game in progress
				if (currentGame.getGameEndMillis() == -1)
				{
					LeftBid bid = new LeftBid();
					Player player = new Player(playerNumber, EntropyUtil.getColourForPlayerNumber(playerNumber));
					player.setName(username);
					bid.setPlayer(player);
					
					BidHistory history = currentGame.getCurrentBidHistory();
					history.addBidForPlayer(playerNumber, bid);
					
					int roundNumber = currentGame.getRoundNumber();
					if (roundNumber > 1)
					{
						StatisticsUtil.recordGamePlayed(username, capacity, settings.getMode());
					}
					
					//Moved this into here as otherwise we set it to 0 incorrectly and a person ends up with no cards!
					HandDetails details = currentGame.getCurrentRoundDetails();
					ConcurrentHashMap<Integer, Integer> hmHandSizeByPlayerNumber = details.getHandSizes();
					hmHandSizeByPlayerNumber.put(playerNumber, 0);
				}
				
				int playerSize = hmPlayerByPlayerNumber.size();
				if (playerSize == 1)
				{
					int remainingPlayerNumber = hmPlayerByPlayerNumber.getOnlyKey();
					finishCurrentGame(remainingPlayerNumber);
				}
				else if (playerSize == 0)
				{
					resetCurrentPlayers(fireLobbyChanged);
					clearChatIfEmpty();
				}
			}
		}
	}
	
	private void notifyAllPlayersOfPlayerChange(String userToExclude, boolean blocking)
	{
		String notification = XmlBuilderServer.getPlayerNotification(this);
		notifyAllUsersViaGameSocket(notification, userToExclude, blocking);
	}
	
	private void notifyAllUsersViaGameSocket(String notification, String userToExclude, boolean blocking)
	{
		HashSet<String> usersToNotify = getAllUsersInRoom();
		if (userToExclude != null)
		{
			usersToNotify.remove(userToExclude);
		}
		
		List<UserConnection> uscs = ServerGlobals.INSTANCE.getUscStore().getAllForNames(usersToNotify);
		server.sendViaNotificationSocket(uscs, notification, XmlConstants.SOCKET_NAME_GAME, blocking);
	}
	
	private void finishCurrentGame(int winningPlayer)
	{
		int roundNumber = currentGame.getRoundNumber();
		String winningUsername = hmPlayerByPlayerNumber.get(winningPlayer);
		saveStatistics(winningUsername, roundNumber);
		resetCurrentPlayers();
		currentGame.setWinningPlayer(winningPlayer);
		
		if (roundNumber > 1)
		{
			OnlineMessage om = new OnlineMessage("black", winningUsername + " won!", "Game");
			addToChatHistoryAndNotifyUsers(om);
		}
		
		initialiseGame();
		
		//Notify everyone that the game is over now we've finished setting up
		String notification = XmlBuilderServer.factoryGameOverNotification(this, winningPlayer);
		notifyAllUsersViaGameSocket(notification, null, false);
	}
	
	public void resetCurrentPlayers()
	{
		resetCurrentPlayers(true);
	}
	public void resetCurrentPlayers(boolean fireLobbyChanged)
	{
		currentPlayers.clear();
		for (int i=0; i<capacity; i++)
		{
			String username = hmPlayerByPlayerNumber.get(i);
			if (username != null)
			{
				currentPlayers.add(username);
			}
		}
		
		hmFormerPlayerByPlayerNumber.clear();
		if (fireLobbyChanged)
		{
			ServerGlobals.lobbyService.lobbyChanged();
		}
	}
	
	
	public void addToObservers(String username)
	{
		synchronized (this)
		{
			if (!observers.contains(username))
			{
				observers.add(username);
			}
			
			removePlayer(username, false);
			ServerGlobals.lobbyService.lobbyChanged();
		}
	}
	public void removeFromObservers(String username)
	{
		boolean removed = observers.remove(username);
		if (removed)
		{
			ServerGlobals.lobbyService.lobbyChanged();
		}
		
		clearChatIfEmpty();
	}
	
	public void initialiseGame()
	{
		try
		{
			String gameId = "G" + System.currentTimeMillis();
			if (currentGame != null)
			{
				previousGame = currentGame.factoryCopy();
			}
			
			currentGame = new GameWrapper(gameId);
			
			HandDetails details = new HandDetails();
			ExtendedConcurrentHashMap<Integer, Integer> hmHandSizeByPlayerNumber = new ExtendedConcurrentHashMap<>();
			for (int i=0; i<capacity; i++)
			{
				hmHandSizeByPlayerNumber.put(i, Integer.valueOf(5));
			}
			
			ConcurrentHashMap<Integer, String[]> hmHandByPlayerNumber = dealHandsHashMap(hmHandSizeByPlayerNumber);
			
			details.setHands(hmHandByPlayerNumber);
			details.setHandSizes(hmHandSizeByPlayerNumber);
			currentGame.setDetailsForRound(1, details);
	
			Random rand = new Random();
			int personToStart = rand.nextInt(capacity);
			
			BidHistory history = new BidHistory();
			history.setPersonToStart(personToStart);
			currentGame.setBidHistoryForRound(1, history);
		}
		catch (Throwable t)
		{
			Debug.stackTrace(t);
		}
	}
	
	public void handleChallenge(String gameId, int roundNumber, int playerNumber, int challengedNumber, Bid bid) 
	{
		GameWrapper game = getGameForId(gameId);
		HandDetails details = game.getDetailsForRound(roundNumber);
		ConcurrentHashMap<Integer, String[]> hmHandByPlayerNumber = details.getHands();
		if (bid.isOverbid(hmHandByPlayerNumber, settings.getJokerValue()))
		{
			//bidder loses
			setUpNextRound(challengedNumber);
		}
		else
		{
			//challenger loses
			setUpNextRound(playerNumber);
		}
	}
	
	public void handleIllegal(String gameId, int roundNumber, int playerNumber, int bidderNumber, Bid bid) 
	{
		GameWrapper game = getGameForId(gameId);
		HandDetails details = game.getDetailsForRound(roundNumber);
		ConcurrentHashMap<Integer, String[]> hmHandByPlayerNumber = details.getHands();
		if (bid.isPerfect(hmHandByPlayerNumber, settings.getJokerValue(), settings.getIncludeMoons(), settings.getIncludeStars()))
		{
			setUpNextRound(bidderNumber);
		}
		else
		{
			setUpNextRound(playerNumber);
		}
	}
	
	public void setUpNextRound(int losingPlayerNumber)
	{
		int currentRoundNumber = currentGame.getRoundNumber();
		
		HandDetails nextRoundDetails = currentGame.getDetailsForRound(currentRoundNumber + 1);
		if (nextRoundDetails != null)
		{
			Debug.stackTrace("Trying to set up next round but it's not null. Room " + name);
			return;
		}
		
		HandDetails currentRoundDetails = currentGame.getDetailsForRound(currentRoundNumber);
		ExtendedConcurrentHashMap<Integer, Integer> hmHandSizeByPlayerNumber = currentRoundDetails.getHandSizes();
		Integer handSize = hmHandSizeByPlayerNumber.get(losingPlayerNumber);
		int newHandSize = Math.max(0, handSize.intValue() - 1);
		
		ExtendedConcurrentHashMap<Integer, Integer> hmHandSizeByPlayerNumberForNextRound = hmHandSizeByPlayerNumber.factoryCopy();
		hmHandSizeByPlayerNumberForNextRound.put(losingPlayerNumber, Integer.valueOf(newHandSize));
		
		int potentialWinner = getWinningPlayer(hmHandSizeByPlayerNumberForNextRound);
		if (potentialWinner > -1)
		{
			finishCurrentGame(potentialWinner);
		}
		else
		{
			nextRoundDetails = new HandDetails();
			nextRoundDetails.setHandSizes(hmHandSizeByPlayerNumberForNextRound);
			
			ConcurrentHashMap<Integer, String[]> hmHandByPlayerNumber = dealHandsHashMap(hmHandSizeByPlayerNumberForNextRound);
			
			nextRoundDetails.setHands(hmHandByPlayerNumber);
			currentGame.setDetailsForRound(currentRoundNumber + 1, nextRoundDetails);
			
			BidHistory history = new BidHistory();
			history.setPersonToStart(losingPlayerNumber);
			currentGame.setBidHistoryForRound(currentRoundNumber + 1, history);
			
			currentGame.setRoundNumber(currentRoundNumber + 1);
			
			String newRoundNotification = XmlBuilderServer.factoryNewRoundNotification(this, nextRoundDetails, losingPlayerNumber);
			notifyAllUsersViaGameSocket(newRoundNotification, null, false);
		}
	}
	
	private ConcurrentHashMap<Integer, String[]> dealHandsHashMap(ExtendedConcurrentHashMap<Integer, Integer> hmHandSizeByPlayerNumber)
	{
		ConcurrentHashMap<Integer, String[]> hmHandByPlayerNumber = new ConcurrentHashMap<>();
		
		long seed = server.generateSeed();
		List<String> deck = CardsUtil.createAndShuffleDeck(true, settings.getJokerQuantity(),
				settings.getIncludeMoons(), settings.getIncludeStars(), settings.getNegativeJacks(), seed);
		for (int i=0; i<capacity; i++)
		{
			int size = hmHandSizeByPlayerNumber.get(i);
			String hand[] = new String[size];
			for (int j=0; j<size; j++)
			{
				hand[j] = deck.remove(j);
			}
			
			hmHandByPlayerNumber.put(i, hand);
		}
		
		return hmHandByPlayerNumber;
	}
	
	private int getWinningPlayer(ConcurrentHashMap<Integer, Integer> hmHandSizeByPlayerNumber)
	{
		int activePlayers = 0;
		int potentialWinner = 0;
		
		for (int i=0; i<capacity; i++)
		{
			int handSize = hmHandSizeByPlayerNumber.get(i);
			if (handSize > 0)
			{
				activePlayers++;
				potentialWinner = i;
			}
		}
		
		if (activePlayers > 1)
		{
			return -1;
		}
		else
		{
			return potentialWinner;
		}
	}
	
	private void saveStatistics(String winningPlayer, int roundNumber)
	{
		//don't save any statistics if the game didn't go beyond the first round
		if (roundNumber == 1)
		{
			return;
		}
		
		StatisticsUtil.recordWin(winningPlayer, capacity, settings.getMode());
		
		int size = currentPlayers.size();
		for (int i=0; i<size; i++)
		{
			String player = currentPlayers.get(i);
			if (hmPlayerByPlayerNumber.containsValue(player))
			{
				//They didn't leave, so record a game played
				StatisticsUtil.recordGamePlayed(player, capacity, settings.getMode());
				
				//Push out a stats notification
				Document statsNotification = XmlBuilderServer.factoryStatisticsNotification(player);
				String notificationString = XmlUtil.getStringFromDocument(statsNotification);
				UserConnection usc = ServerGlobals.INSTANCE.getUscStore().findForName(player);
				usc.sendNotificationInWorkerPool(notificationString, server, XmlConstants.SOCKET_NAME_LOBBY, null);
			}
		}
	}
	
	public long getCurrentGameDuration()
	{
		long gameStart = currentGame.getGameStartMillis();
		if (gameStart == -1
		  || currentGame.getRoundNumber() == 1)
		{
			return 0;
		}
		
		long currentTime = System.currentTimeMillis();
		return (currentTime - gameStart);
	}

	/**
	 * Returns a HashSet since it's possible for a player to be present as a player AND an observer.
	 * This occurs if they've left but the game is still going - we keep the reference as a player so
	 * others can't take the seat. They obviously then have the option to join as an observer.
	 */
	public HashSet<String> getAllUsersInRoom()
	{
		ArrayList<String> ret = getCurrentPlayers();
		ret.addAll(getObservers());

		HashSet<String> hs = new HashSet<>(ret);
		return hs;
	}
	
	/**
	 * Get/sets
	 */
	public String getName() {
		return name;
	}
	public int getCapacity() {
		return capacity;
	}
	public GameSettings getSettings() {
		return settings;
	}
	public boolean isGameInProgress()
	{
		//This no longer works
		//return currentGame != null;
		//return currentPlayers.size() == capacity && (waitingForPlayerToSeeResult == null);
		return currentPlayers.size() == capacity;
	}
	public boolean getIsCopy()
	{
		return isCopy;
	}
	public void setIsCopy(boolean isCopy)
	{
		this.isCopy = isCopy;
	}
	public ArrayList<String> getCurrentPlayers()
	{
		return new ArrayList<>(currentPlayers);
	}
	public int getCurrentPlayerCount()
	{
		return currentPlayers.size();
	}
	public ArrayList<String> getObservers()
	{
		return new ArrayList<>(observers);
	}
	public int getObserverCount()
	{
		return observers.size();
	}

	/**
	 * HashMap gets/sets
	 */
	public String getPlayer(int playerNumber)
	{
		return hmPlayerByPlayerNumber.get(playerNumber);
	}
	public String getFormerPlayer(int playerNumber)
	{
		return hmFormerPlayerByPlayerNumber.get(playerNumber);
	}
	public GameWrapper getGameForId(String gameId)
	{
		String currentId = currentGame.getGameId();
		if (gameId.equals(currentId))
		{
			return currentGame;
		}
		
		String previousId = previousGame.getGameId();
		if (gameId.equals(previousId))
		{
			return previousGame;
		}

		throw new RuntimeException("Got a null game for room " + name + " and gameId " + gameId);
	}
	public GameWrapper getNextGameForId(String previousGameIdFromClient)
	{
		if (previousGameIdFromClient.isEmpty())
		{
			return currentGame;
		}
		
		if (previousGame == null)
		{
			Debug.append("Tried to get next game for gameId " + previousGameIdFromClient + " but previous game was null.");
			Debug.appendWithoutDate("Current: " + currentGame.getGameId());
			return currentGame;
		}
		
		String previousGameId = previousGame.getGameId();
		if (previousGameId.equals(previousGameIdFromClient))
		{
			return currentGame;
		}
		
		Debug.append("Tried to get next game for gameId " + previousGameIdFromClient + " but this didn't match my previous game.");
		Debug.appendWithoutDate("Previous: " + previousGame.getGameId());
		Debug.appendWithoutDate("Current: " + currentGame.getGameId());
		return currentGame;
	}
	
	
	public Bid getLastBidForPlayer(int playerNumber, int roundNumber)
	{
		if (currentGame == null
		  || playerNumber == -1)
		{
			return null;
		}
		
		BidHistory history = currentGame.getBidHistoryForRound(roundNumber);
		return history.getLastBidForPlayer(playerNumber);
	}
	public boolean addBidForPlayer(String gameId, int playerNumber, int roundNumber, Bid newBid)
	{
		GameWrapper game = getGameForId(gameId);
		
		BidHistory history = game.getBidHistoryForRound(roundNumber);
		boolean added = history.addBidForPlayer(playerNumber, newBid);
		
		if (added)
		{
			//Notify all other capacity
			String bidNotification = XmlBuilderServer.getBidNotification(name, playerNumber, newBid);
			notifyAllUsersViaGameSocket(bidNotification, null, false);
		}
		
		return added;
	}
	
	public void addToChatHistoryAndNotifyUsers(OnlineMessage message)
	{
		if (isEmpty())
		{
			return;
		}
		
		chatHistory.add(message);
		
		String chatMessage = XmlBuilderServer.getChatNotification(name, message);
		HashSet<String> users = getAllUsersInRoom();
		List<UserConnection> uscs = ServerGlobals.INSTANCE.getUscStore().getAllForNames(users);
		server.sendViaNotificationSocket(uscs, chatMessage, XmlConstants.SOCKET_NAME_CHAT, false);
	}
	public ArrayList<OnlineMessage> getChatHistory()
	{
		return chatHistory;
	}
	
	@Override
	public String toString() 
	{
		return name;
	}
}

File: server/src/main/kotlin/room/Room.kt
package room

import auth.UserConnection
import game.GameSettings
import http.dto.JoinRoomResponse
import http.dto.OnlineMessage
import http.dto.RoomStateResponse
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.max
import `object`.Bid
import `object`.BidHistory
import `object`.GameWrapper
import `object`.HandDetails
import `object`.LeftBid
import `object`.Player
import store.IHasId
import util.CardsUtil
import util.EntropyUtil
import util.ServerGlobals
import util.ServerGlobals.roomStore
import util.ServerGlobals.uscStore
import util.XmlBuilderServer
import util.XmlConstants
import utils.CoreGlobals.logger

data class Room(
    override val id: UUID,
    private val baseName: String,
    val settings: GameSettings,
    val capacity: Int,
    private val index: Int = 1,
) : IHasId<UUID> {
    private val hmPlayerByPlayerNumber = ConcurrentHashMap<Int, String>()
    private val hmFormerPlayerByPlayerNumber: ConcurrentHashMap<Int, String> = ConcurrentHashMap()
    val chatHistory: MutableList<OnlineMessage> = mutableListOf()
    private val currentPlayers: MutableList<String> = mutableListOf()
    private val observers: MutableList<String> = mutableListOf()
    private var previousGame: GameWrapper? = null
    private var currentGame: GameWrapper = initialiseGame()

    val name: String
        get() = "$baseName $index"

    val isCopy: Boolean
        get() = index > 1

    val isFull: Boolean
        get() = currentPlayers.size == capacity

    private val isEmpty: Boolean
        get() = (currentPlayers.isEmpty() && observers.isEmpty())

    private fun clearChatIfEmpty() {
        synchronized(this) {
            if (!isEmpty) {
                return
            }
            chatHistory.clear()
        }
    }

    fun getColourForPlayer(playerName: String): String {
        val playerNumber =
            hmPlayerByPlayerNumber.filter { it.value == playerName }.keys.firstOrNull()

        return if (playerNumber != null) {
            EntropyUtil.getColourForPlayerNumber(playerNumber)
        } else "gray"
    }

    fun attemptToSitDown(username: String, playerNumber: Int): Int? {
        synchronized(this) {
            val existingUsername: String? = hmPlayerByPlayerNumber[playerNumber]
            if (existingUsername != null) {
                logger.info(
                    "seatTaken",
                    "$username tried to join $name as player $playerNumber but seat was taken by $existingUsername",
                )
                return null
            }

            if (hmPlayerByPlayerNumber.containsValue(username)) {
                logger.info("doubleJoin", "$username tried to join $name twice!")
                return null
            }

            currentPlayers.add(username)
            hmPlayerByPlayerNumber[playerNumber] = username
            observers.remove(username)

            if (isFull) {
                roomStore.addCopy(this)
            }

            notifyAllPlayersOfPlayerChange(username, false)
            ServerGlobals.lobbyService.lobbyChanged()
            return playerNumber
        }
    }

    fun removePlayer(username: String, fireLobbyChanged: Boolean) {
        for (playerNumber in 0..<capacity) {
            val user: String? = hmPlayerByPlayerNumber[playerNumber]
            if ((user != null && (username == user))) {
                hmPlayerByPlayerNumber.remove(playerNumber)
                hmFormerPlayerByPlayerNumber[playerNumber] = user

                // Notify everyone in the room that this player has left. Block on this.
                notifyAllPlayersOfPlayerChange(username, true)

                // The game has not started
                if (currentGame.gameStartMillis == -1L) {
                    // Unset the countdown if it's going, reset current capacity and get out of this
                    // madness
                    currentGame.setCountdownStartMillis(-1)
                    resetCurrentPlayers(fireLobbyChanged)
                    return
                }

                // There is a game in progress
                if (currentGame.gameEndMillis == -1L) {
                    val bid = LeftBid()
                    val player =
                        Player(playerNumber, EntropyUtil.getColourForPlayerNumber(playerNumber))
                    player.name = username
                    bid.player = player

                    val history: BidHistory = currentGame.currentBidHistory
                    history.addBidForPlayer(playerNumber, bid)

                    // TODO - Hand size thing
                    // Moved this into here as otherwise we set it to 0 incorrectly and a person
                    // ends up with no cards!
                    //                    val details: HandDetails = currentGame.currentRoundDetails
                    //                    val hmHandSizeByPlayerNumber = details.handSizes
                    //                    hmHandSizeByPlayerNumber[playerNumber] = 0
                }

                val playerSize: Int = hmPlayerByPlayerNumber.size
                if (playerSize == 1) {
                    val remainingPlayerNumber: Int = hmPlayerByPlayerNumber.keys.first()
                    finishCurrentGame(remainingPlayerNumber)
                } else if (playerSize == 0) {
                    resetCurrentPlayers(fireLobbyChanged)
                    clearChatIfEmpty()
                }
            }
        }
    }

    private fun notifyAllPlayersOfPlayerChange(userToExclude: String, blocking: Boolean) {
        val notification: String = XmlBuilderServer.getPlayerNotification(this)
        notifyAllUsersViaGameSocket(notification, userToExclude, blocking)
    }

    private fun notifyAllUsersViaGameSocket(
        notification: String,
        userToExclude: String?,
        blocking: Boolean,
    ) {
        var usersToNotify = allUsersInRoom
        if (userToExclude != null) {
            usersToNotify = usersToNotify.minus(userToExclude)
        }

        val uscs: List<UserConnection> = uscStore.getAllForNames(usersToNotify)
        ServerGlobals.server.sendViaNotificationSocket(
            uscs,
            notification,
            XmlConstants.SOCKET_NAME_GAME,
            blocking,
        )
    }

    private fun finishCurrentGame(winningPlayer: Int) {
        val roundNumber: Int = currentGame.roundNumber
        val winningUsername: String? = hmPlayerByPlayerNumber[winningPlayer]
        resetCurrentPlayers()
        currentGame.winningPlayer = winningPlayer

        if (roundNumber > 1) {
            val om = OnlineMessage("black", "$winningUsername won!", "Game")
            addToChatHistory(om)
        }

        previousGame = currentGame.factoryCopy()
        currentGame = initialiseGame()

        // Notify everyone that the game is over now we've finished setting up
        val notification: String = XmlBuilderServer.factoryGameOverNotification(this, winningPlayer)
        notifyAllUsersViaGameSocket(notification, null, false)
    }

    @JvmOverloads
    fun resetCurrentPlayers(fireLobbyChanged: Boolean = true) {
        currentPlayers.clear()
        for (i in 0..<capacity) {
            val username: String? = hmPlayerByPlayerNumber[i]
            if (username != null) {
                currentPlayers.add(username)
            }
        }

        hmFormerPlayerByPlayerNumber.clear()
        if (fireLobbyChanged) {
            ServerGlobals.lobbyService.lobbyChanged()
        }
    }

    fun addToObservers(username: String) {
        synchronized(this) {
            if (!observers.contains(username)) {
                observers.add(username)
            }
            removePlayer(username, false)
            ServerGlobals.lobbyService.lobbyChanged()
        }
    }

    fun removeFromObservers(username: String) {
        val removed: Boolean = observers.remove(username)
        if (removed) {
            ServerGlobals.lobbyService.lobbyChanged()
        }

        clearChatIfEmpty()
    }

    private fun initialiseGame(): GameWrapper {
        val gameId: String = "G" + System.currentTimeMillis()

        val newGame = GameWrapper(gameId)

        val details = HandDetails()
        val handSizes = (0..capacity).associateWith { 5 }

        val hmHandByPlayerNumber = dealHandsHashMap(handSizes)
        details.hands = hmHandByPlayerNumber
        newGame.setDetailsForRound(1, details)

        val personToStart = Random().nextInt(capacity)

        val history = BidHistory()
        history.personToStart = personToStart
        newGame.setBidHistoryForRound(1, history)

        return newGame
    }

    fun handleChallenge(
        gameId: String,
        roundNumber: Int,
        playerNumber: Int,
        challengedNumber: Int,
        bid: Bid,
    ) {
        val game = getGameForId(gameId)
        val details: HandDetails = game.getDetailsForRound(roundNumber)
        val hmHandByPlayerNumber: ConcurrentHashMap<Int, List<String>> = details.hands
        if (bid.isOverbid(hmHandByPlayerNumber, settings.jokerValue)) {
            // bidder loses
            setUpNextRound(challengedNumber)
        } else {
            // challenger loses
            setUpNextRound(playerNumber)
        }
    }

    fun handleIllegal(
        gameId: String,
        roundNumber: Int,
        playerNumber: Int,
        bidderNumber: Int,
        bid: Bid,
    ) {
        val game = getGameForId(gameId)
        val details: HandDetails = game.getDetailsForRound(roundNumber)
        val hmHandByPlayerNumber: ConcurrentHashMap<Int, List<String>> = details.hands
        if (
            bid.isPerfect(
                hmHandByPlayerNumber,
                settings.jokerValue,
                settings.includeMoons,
                settings.includeStars,
            )
        ) {
            setUpNextRound(bidderNumber)
        } else {
            setUpNextRound(playerNumber)
        }
    }

    private fun setUpNextRound(losingPlayerNumber: Int) {
        val currentRoundNumber: Int = currentGame.roundNumber

        var nextRoundDetails: HandDetails? = currentGame.getDetailsForRound(currentRoundNumber + 1)
        if (nextRoundDetails != null) {
            logger.error("doubleRound", "Trying to set up next round but it's not null. Room $name")
            return
        }

        val currentRoundDetails: HandDetails = currentGame.getDetailsForRound(currentRoundNumber)
        val hands = currentRoundDetails.hands

        val handSizesForNextRound =
            hands.mapValues { (playerNumber, hand) ->
                if (playerNumber == losingPlayerNumber) max(0, hand.size - 1) else hand.size
            }

        val potentialWinner: Int = getWinningPlayer(handSizesForNextRound)
        if (potentialWinner > -1) {
            finishCurrentGame(potentialWinner)
        } else {
            nextRoundDetails = HandDetails()

            val hmHandByPlayerNumber: ConcurrentHashMap<Int, List<String>> =
                dealHandsHashMap(handSizesForNextRound)

            nextRoundDetails.hands = hmHandByPlayerNumber
            currentGame.setDetailsForRound(currentRoundNumber + 1, nextRoundDetails)

            val history = BidHistory()
            history.personToStart = losingPlayerNumber
            currentGame.setBidHistoryForRound(currentRoundNumber + 1, history)

            currentGame.roundNumber = currentRoundNumber + 1

            val newRoundNotification: String =
                XmlBuilderServer.factoryNewRoundNotification(
                    this,
                    nextRoundDetails,
                    losingPlayerNumber,
                )
            notifyAllUsersViaGameSocket(newRoundNotification, null, false)
        }
    }

    private fun dealHandsHashMap(
        hmHandSizeByPlayerNumber: Map<Int, Int>
    ): ConcurrentHashMap<Int, List<String>> {
        val hmHandByPlayerNumber: ConcurrentHashMap<Int, List<String>> = ConcurrentHashMap()

        val seed: Long = ServerGlobals.server.generateSeed()
        val deck =
            CardsUtil.createAndShuffleDeck(
                settings.jokerQuantity,
                settings.includeMoons,
                settings.includeStars,
                settings.negativeJacks,
                seed,
            )

        for (i in 0..<capacity) {
            val size: Int = hmHandSizeByPlayerNumber.getValue(i)
            val hand = (0..<size).map { deck.removeAt(0) }
            hmHandByPlayerNumber[i] = hand
        }

        return hmHandByPlayerNumber
    }

    private fun getWinningPlayer(hmHandSizeByPlayerNumber: Map<Int, Int>): Int {
        var activePlayers = 0
        var potentialWinner = 0

        for (i in 0..<capacity) {
            val handSize: Int = hmHandSizeByPlayerNumber.getValue(i)
            if (handSize > 0) {
                activePlayers++
                potentialWinner = i
            }
        }

        return if (activePlayers > 1) {
            -1
        } else {
            potentialWinner
        }
    }

    /**
     * Returns a set since it's possible for a player to be present as a player AND an observer.
     * This occurs if they've left but the game is still going - we keep the reference as a player
     * so others can't take the seat. They obviously then have the option to join as an observer.
     */
    val allUsersInRoom: Set<String>
        get() = (currentPlayers + observers).toSet()

    val isGameInProgress: Boolean
        get() = currentPlayers.size == capacity

    val currentPlayerCount: Int
        get() = currentPlayers.size

    val observerCount: Int
        get() = observers.size

    fun getPlayer(playerNumber: Int): String? {
        return hmPlayerByPlayerNumber[playerNumber]
    }

    private fun getGameForId(gameId: String): GameWrapper {
        if (gameId == currentGame.gameId) {
            return currentGame
        }

        if ((gameId == previousGame?.gameId)) {
            return previousGame!!
        }

        throw RuntimeException("Got a null game for room $name and gameId $gameId")
    }

    fun getNextGameForId(previousGameIdFromClient: String): GameWrapper {
        if (previousGameIdFromClient.isEmpty()) {
            return currentGame
        }

        if (previousGame == null) {
            logger.warn(
                "staleGameId",
                "Tried to get next game for gameId $previousGameIdFromClient but previous game was null. Current: ${currentGame.gameId}",
            )
            return currentGame
        }

        val previousGameId: String = previousGame!!.gameId
        if (previousGameId == previousGameIdFromClient) {
            return currentGame
        }

        logger.warn(
            "staleGameId",
            "Tried to get next game for gameId $previousGameIdFromClient but this did not match. Previous [${previousGame?.gameId}],  Current [${currentGame.gameId}]",
        )

        return currentGame
    }

    fun getLastBidForPlayer(playerNumber: Int, roundNumber: Int): Bid? {
        if (playerNumber == -1) {
            return null
        }

        val history: BidHistory = currentGame.getBidHistoryForRound(roundNumber)
        return history.getLastBidForPlayer(playerNumber)
    }

    fun addBidForPlayer(
        gameId: String,
        playerNumber: Int,
        roundNumber: Int,
        newBid: Bid?,
    ): Boolean {
        val game = getGameForId(gameId)

        val history: BidHistory = game.getBidHistoryForRound(roundNumber)
        val added: Boolean = history.addBidForPlayer(playerNumber, newBid)

        if (added) {
            // Notify all other capacity
            val bidNotification = XmlBuilderServer.getBidNotification(name, playerNumber, newBid)
            notifyAllUsersViaGameSocket(bidNotification, null, false)
        }

        return added
    }

    fun addToChatHistory(message: OnlineMessage) {
        chatHistory.add(message)
    }

    fun buildJoinRoomResponse() =
        JoinRoomResponse(
            chatHistory,
            hmPlayerByPlayerNumber.toMap(),
            hmFormerPlayerByPlayerNumber.toMap(),
        )

    fun buildRoomStateResponse() =
        RoomStateResponse(hmPlayerByPlayerNumber.toMap(), hmFormerPlayerByPlayerNumber.toMap())

    fun makeCopy() = Room(UUID.randomUUID(), baseName, settings, capacity, index + 1)

    override fun toString() = name
}
